php/php-8.2.0/patches/0017-fix-is_writable-by-trial-and-error.diff

```diff
@@ -0,0 +1,104 @@
+From dafe004599782efe80832782b2908b06a40eb9c2 Mon Sep 17 00:00:00 2001
+From: =?UTF-8?q?Jesu=CC=81s=20Gonza=CC=81lez?= <jesusgm@vmware.com>
+Date: Wed, 12 Apr 2023 13:35:19 +0200
+Subject: [PATCH] fix: Emulated implementations for `is_readable()` and
+ `is_writable()` by trial and error
+
+---
+ main/streams/plain_wrapper.c | 72 ++++++++++++++++++++++++++++++++++++
+ 1 file changed, 72 insertions(+)
+
+diff --git a/main/streams/plain_wrapper.c b/main/streams/plain_wrapper.c
+index f72825e1..5aa14768 100644
+--- a/main/streams/plain_wrapper.c
++++ b/main/streams/plain_wrapper.c
+@@ -1214,6 +1214,44 @@ static php_stream *php_plain_files_stream_opener(php_stream_wrapper *wrapper, co
+ 	return php_stream_fopen_rel(path, mode, opened_path, options);
+ }
+ 
++#ifdef WASM_WASI
++int wasi_dir_is_readable(const char* dir)
++{
++	DIR* dd = opendir(dir);
++	if (dd == NULL)
++		return 0;
++	
++	closedir(dd);
++	return 1;
++}
++
++int wasi_dir_is_writable(const char* dir)
++{
++	char path[1024];
++	char test_file[32] = "PHP_WASM_WASI_WRITABLE_TESTING";
++
++	snprintf(path, sizeof(path)-1, "%s/%s", dir, test_file);
++	FILE* fd = fopen(path, "w");
++	if (fd == NULL) 
++		return 0;
++
++	fclose(fd);
++	remove(path);
++	return 1;
```

Comment:
I get that this is the only way to test this to be almost sure -- as there is still room for false positives (the file PHP_WASM_WASI_WRITABLE_TESTING did exist and is writable, but the directory is not), and false negatives, the other way around. I know it seems like edge cases, there might be others I cannot think of right now.

I think this patch is something we should probably carry ourselves on top of PHP, even when all the wasm32-wasi changes are upstreamed, as I don't see feasible to contribute this upstream directly (I might be mistaken though...)

Reply:
Regarding `PHP_WASM_WASI_WRITABLE_TESTING` existing, yes, you are right. I thought about generating a random number and using it as the filename, but then I noticed how PHP uses `/tmp/phpXXXXXX` for temporary files. So, I decided to keep the patch simpler. Also, if for whatever reason the `remove()` instruction is not executed, we don't have an accumulation of files.

For the edge cases not covered, I agree 100%. The fact that this patch seems to work with Drupal, doesn't really mean it will work with other apps. We would need to review new cases but I also think new corner cases not covered here will raise.

About the patch, this reminds me the `mmap` emulation in wasi-libc. Our implementation might be valid for most scenarios but in others maybe PHP's `is_readable()` and `is_writable()` should be avoided since WASI preview1 simply does not cover this case. Hopefully, WASI preview2 will make this patch obsolete! :-)